core/matrix/batch_dense.cpp

```diff
@@ -0,0 +1,251 @@
+    // Otherwise, it results in a compile error.
+    return (*other).create_with_same_config();
+}
+
+
+template <typename ValueType>
+std::unique_ptr<Dense<ValueType>> Dense<ValueType>::create_with_same_config()
+    const
+{
+    return Dense<ValueType>::create(this->get_executor(), this->get_size());
+}
+
+
+template <typename ValueType>
+std::unique_ptr<const Dense<ValueType>> Dense<ValueType>::create_const(
+    std::shared_ptr<const Executor> exec, const batch_dim<2>& sizes,
+    gko::detail::const_array_view<ValueType>&& values)
+{
+    // cast const-ness away, but return a const object afterwards,
+    // so we can ensure that no modifications take place.
+    return std::unique_ptr<const Dense>(new Dense{
+        exec, sizes, gko::detail::array_const_cast(std::move(values))});
+}
+
+
+inline const batch_dim<2> get_col_sizes(const batch_dim<2>& sizes)
+{
+    return batch_dim<2>(sizes.get_num_batch_items(),
+                        dim<2>(1, sizes.get_common_size()[1]));
+}
+
+
+template <typename ValueType>
+Dense<ValueType>::Dense(std::shared_ptr<const Executor> exec,
+                        const batch_dim<2>& size)
+    : EnableBatchLinOp<Dense<ValueType>>(exec, size),
+      values_(exec, compute_num_elems(size))
+{}
+
+
+template <typename ValueType>
+void Dense<ValueType>::apply_impl(const MultiVector<ValueType>* b,
+                                  MultiVector<ValueType>* x) const
+{
+    GKO_ASSERT_EQ(b->get_num_batch_items(), this->get_num_batch_items());
+    GKO_ASSERT_EQ(this->get_num_batch_items(), x->get_num_batch_items());
+
+    GKO_ASSERT_CONFORMANT(this->get_common_size(), b->get_common_size());
+    GKO_ASSERT_EQUAL_ROWS(this->get_common_size(), x->get_common_size());
+    GKO_ASSERT_EQUAL_COLS(b->get_common_size(), x->get_common_size());
+    this->get_executor()->run(dense::make_simple_apply(this, b, x));
+}
+
+
+template <typename ValueType>
+void Dense<ValueType>::apply_impl(const MultiVector<ValueType>* alpha,
+                                  const MultiVector<ValueType>* b,
+                                  const MultiVector<ValueType>* beta,
+                                  MultiVector<ValueType>* x) const
+{
+    GKO_ASSERT_EQ(b->get_num_batch_items(), this->get_num_batch_items());
+    GKO_ASSERT_EQ(this->get_num_batch_items(), x->get_num_batch_items());
+
+    GKO_ASSERT_CONFORMANT(this->get_common_size(), b->get_common_size());
+    GKO_ASSERT_EQUAL_ROWS(this->get_common_size(), x->get_common_size());
+    GKO_ASSERT_EQUAL_COLS(b->get_common_size(), x->get_common_size());
+    GKO_ASSERT_EQUAL_DIMENSIONS(alpha->get_common_size(), gko::dim<2>(1, 1));
+    GKO_ASSERT_EQUAL_DIMENSIONS(beta->get_common_size(), gko::dim<2>(1, 1));
+    this->get_executor()->run(
+        dense::make_advanced_apply(alpha, this, b, beta, x));
+}
+
+
+template <typename ValueType>
+void Dense<ValueType>::convert_to(
+    Dense<next_precision<ValueType>>* result) const
+{
+    result->values_ = this->values_;
```

Comment:
```suggestion
    result->values_ = array<next_precision<ValueType>(result->get_executor, this->values_);
```

Reply:
array has a overloaded copy operator with the necessary checks, so it is cleaner to use that.

Reply:
right, I somehow forgot about that.

Reply:
side-note: This can mess with submatrices/strided storage, since it also overwrites padding if the target is already allocated. I'm not sure if we do that correctly in `matrix::Dense`, but it should be mentioned.